### PR TITLE
Loki: implement decolorize filter

### LINKED ISSUE #7601: Make it easier to parse colorized logs
**Is your feature request related to a problem? Please describe.**

It's hard to parse logs (i.e. with `pattern`) if they are colored. Consider a (real) log line from [Strapi](https://strapi.io/):

`[2022-11-04 22:17:57.811] http: GET /_health (0 ms) 204`

where `http` is green.

To extract data from it, we would need this query (notice the weird bits: `[32m`, `[39m`):
```
{job="strapi/strapi"} | pattern `[<date> <time>] [32mhttp[39m: <method> <path> (<ms> ms) <status>`
```

**Describe the solution you'd like**
A new filter (i.e. `decolorize`) allowing to get rid of all colors (ANSI codes). Pipeline from above would then look like this:

```
{job="strapi/strapi"} | decolorize | pattern `[<date> <time>] http: <method> <path> (<ms> ms) <status>`
```

**Describe alternatives you've considered**

One alternative is to disable colors on the producer side.

### PR DESCRIPTION
**Pull request by vmax**
**What this PR does / why we need it**:
Implement `decolorize` filter that get rids of ANSI color codes in the log message.

**Which issue(s) this PR fixes**:
Fixes #7601

**Special notes for your reviewer**:

It's my first time writing Golang, so please let me know if I missed anything obvious 🙂 

**Checklist**
- [x] Reviewed the `CONTRIBUTING.md` guide
- [x] Documentation added
- [x] Tests updated
- [x] `CHANGELOG.md` updated
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

### CODE REVIEW

**Review by dannykopping — COMMENTED** [review]
I love this idea @vmax! Can you please include some unit tests?

**Comment by dannykopping** [comment]
We can do this in a separate PR, but it would also be great to have this as a promtail stage

**Comment by vmax** [comment]
@dannykopping I've implemented a unit test for `Decolorizer` in https://github.com/grafana/loki/pull/7602/commits/c748326259bc921acb0015a414c73f4ed1923150; modified tests for [parser](https://github.com/grafana/loki/pull/7602/commits/a9d4298a100bf9e643f042b171a1901eb75d47b0) and for [lexer](https://github.com/grafana/loki/pull/7602/commits/55bbf2929f32bf5eec84366bd89142ff12f88406)

**Comment by vmax** [comment]
@dannykopping I've also added `decolorize` [pipeline](https://github.com/grafana/loki/pull/7602/commits/f0be8791832ee6aa740881be6647a26fb0a1dde2) and [tests for it](https://github.com/grafana/loki/pull/7602/commits/c309f2d1e547166d841aeb530d5df41cf4524356)

**Comment by vmax** [comment]
@dannykopping I've amended the commit to include:
- tests for `Decolorizer`
- updated tests for parser and lexer
- `decolorize` stage for Promtail and tests for it

File: CHANGELOG.md [hunk 1]
```diff
@@ -24,6 +24,7 @@
 * [7264](https://github.com/grafana/loki/pull/7264) **bboreham**: Chunks: decode varints directly from byte buffer, for speed.
 * [7263](https://github.com/grafana/loki/pull/7263) **bboreham**: Dependencies: klauspost/compress package to v1.15.11; improves performance.
 * [7270](https://github.com/grafana/loki/pull/7270) **wilfriedroset**: Add support for `username` to redis cache configuration.
+* [7602](https://github.com/grafana/loki/pull/7602) **vmax**: Add decolorize filter to easily parse colored logs.
```
Comment (vlad-diachenko) [inline CHANGELOG.md:27]: could you please move your changelog entries to new sections from https://github.com/grafana/loki/pull/7606

you added your entries to Loki 2.7.0 changelog.

**Review by vlad-diachenko — CHANGES_REQUESTED** [review]
Looks great. thanks a lot for your contribution !!!
it would be awesome if you could also add a section to `docs/sources/logql/log_queries.md` to add documentation on how it works. 
after the merge, the doc will be automatically published to Loki docs page here https://grafana.com/docs/loki/next/logql/log_queries/

File: pkg/logql/syntax/lex.go [hunk 2]
```diff
@@ -66,8 +66,9 @@ var tokens = map[string]int{
 	OpParserTypePattern: PATTERN,
 
 	// fmt
-	OpFmtLabel: LABEL_FMT,
-	OpFmtLine:  LINE_FMT,
+	OpFmtLabel:   LABEL_FMT,
+	OpFmtLine:    LINE_FMT,
+	OpDecolorize: DECOLORIZE,
```
[the remark this answers is not in the document]
Comment (vmax) [reply pkg/logql/syntax/lex.go:71]: @vlad-diachenko indeed! addressed in https://github.com/grafana/loki/pull/7602/commits/f5dabc955ff1ef856d55c35719f808fe3d6bc139

File: CHANGELOG.md (hunk 1 above)
Comment (vmax) [reply CHANGELOG.md]: @vlad-diachenko rebased & moved to `Unreleased` section in https://github.com/grafana/loki/pull/7602/commits/bb901e4691a015976ab37aa33bdb7adad41cac7d

**Review by vlad-diachenko — APPROVED** [review]
Looks brilliant 💎
